File: src/com/easy/AverageOfLevelsInBinaryTree_637.java
/**
 * 
 */
package com.easy;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class AverageOfLevelsInBinaryTree_637 {

    // Use BFS
    // Time Complexity: O(n)
    // Space Complexity: O(h)
    // Runtime: 3 ms, faster than 25.77%
    // Memory Usage: 41.7 MB, less than 42.96%
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> mAnswer = new ArrayList<>();
        Queue<TreeNode> mQueueCurrent = new LinkedList<>();
        Queue<TreeNode> mQueueNext = new LinkedList<>();
        mQueueCurrent.add(root);

        while (!mQueueCurrent.isEmpty()) {
            double mSum = 0;
            for (TreeNode mNode : mQueueCurrent) {
                mSum += mNode.val;
                if (mNode.left != null) {
                    mQueueNext.add(mNode.left);
                }
                if (mNode.right != null) {
                    mQueueNext.add(mNode.right);
                }
            }
            mAnswer.add(mSum / mQueueCurrent.size());

            mQueueCurrent = mQueueNext;
            mQueueNext = new LinkedList<>();

        }

        return mAnswer;
    }

}

File: Golang_src/easy/averageOfLevelsInBinaryTree_637.go
package easy

// TreeNode for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Use BFS:
// Time Complexity: O(n)
// Space Complexity:O(h)
// Runtime: 8 ms, faster than 90.24%
// Memory Usage: 6.2 MB, less than 13.41%
func averageOfLevels(root *TreeNode) []float64 {
	result := []float64{}

	if root == nil {
		return result
	}

	queue := []*TreeNode{root}

	for len(queue) > 0 {
		size := len(queue)
		var sum float64 = 0

		for _, node := range queue {
			sum += float64(node.Val)
			queue = queue[1:]

			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		result = append(result, sum/float64(size))
	}

	return result
}
